File: WidgetLayoutCalculator.cs
namespace ClaudeUsageTray;

internal readonly record struct WidgetLayoutRequest(
    WidgetDensity Density,
    WidgetLayout Layout,
    WidgetVisualTheme Theme,
    bool ShowClaude,
    bool ShowCodex,
    bool ShowProgressBars,
    bool InTaskbar = false,
    double TaskbarBandHeight = WidgetLayoutCalculator.DefaultTaskbarBandHeight);

/// <summary>
/// Geometry of the in-taskbar overlay. Every length is in DIPs and is applied
/// by UsageWidgetWindow to the TaskbarPanel cells without further arithmetic.
/// </summary>
internal readonly record struct TaskbarLayoutMetrics(
    double CellWidth,
    double CellGap,
    double ProviderGap,
    double PaddingX,
    double PaddingY,
    double LabelLineHeight,
    double ValueLineHeight,
    double ProgressHeight,
    double ProgressTopMargin,
    double ProgressSideMargin,
    bool ProgressVisible,
    double AnchorGap);

internal readonly record struct WidgetLayoutMetrics(
    double Width,
    double Height,
    double CompactGap,
    double ComfortableGap,
    double ProviderTop,
    double ProviderGap,
    double SmallCodexMarginLeft,
    double SmallClaudeMarginTop,
    WidgetLayout EffectiveLayout,
    TaskbarLayoutMetrics? Taskbar = null);

/// <summary>
/// Calculates widget geometry without touching WPF controls so every service and
/// density combination shares one auditable set of sizing rules.
/// </summary>
internal static class WidgetLayoutCalculator
{
    internal const double DefaultTaskbarBandHeight = 48;
    internal const double MinimumTaskbarBandHeight = 32;

    // In-taskbar overlay geometry. It ignores density, row layout and visual
    // theme so the overlay keeps one shape inside every supported band height.
    private const double TaskbarCellWidth = 38;
    private const double TaskbarCellGap = 4;
    private const double TaskbarProviderGap = 8;
    private const double TaskbarPaddingX = 6;
    private const double TaskbarPaddingY = 2;
    private const double TaskbarLabelLineHeight = 12;
    private const double TaskbarValueLineHeight = 14;
    private const double TaskbarProgressHeight = 3;
    private const double TaskbarProgressTopMargin = 2;
    private const double TaskbarProgressSideMargin = 3;
    private const double TaskbarAnchorGap = 4;
    private const double TaskbarClippingGuard = 1;

    internal static WidgetLayoutMetrics Calculate(WidgetLayoutRequest request)
    {
        var small = request.Density == WidgetDensity.Small;
        var comfortable = request.Density == WidgetDensity.Comfortable;
        var compact = request.Density == WidgetDensity.Compact;
        var themed = ThemeManager.UsesThemedChrome(request.Theme);
        var providerCount = (request.ShowClaude ? 1 : 0) + (request.ShowCodex ? 1 : 0);
        // TwoRows separates Codex and Claude. With fewer than two visible
        // providers it must be visually identical to SingleRow. The in-taskbar
        // overlay is always a single row.
        var effectiveLayout = !request.InTaskbar && request.Layout == WidgetLayout.TwoRows && providerCount == 2
            ? WidgetLayout.TwoRows
            : WidgetLayout.SingleRow;
        var singleRow = effectiveLayout == WidgetLayout.SingleRow;

        var compactGap = small ? 6 : request.Theme switch
        {
            WidgetVisualTheme.RetroNight => 6,
            WidgetVisualTheme.FluentGlass => 8,
            WidgetVisualTheme.TerminalMono => 5,
            WidgetVisualTheme.Orbit => 8,
            WidgetVisualTheme.PaperInk => 12,
            _ => 10
        };
        var comfortableGap = request.Theme switch
        {
            WidgetVisualTheme.TerminalMono => 8,
            WidgetVisualTheme.PaperInk => 20,
            _ => 14
        };
        const double providerTop = 0;
        var providerGap = compact ? 10 : 12;
        var codexMarginLeft = request.ShowClaude && request.ShowCodex && singleRow ? 8 : 0;
        var claudeMarginTop = request.ShowClaude && request.ShowCodex && !singleRow ? 8 : 0;

        if (request.InTaskbar)
        {
            var (taskbarWidth, taskbarHeight, taskbar) = CalculateTaskbarSize(request);
            return new WidgetLayoutMetrics(taskbarWidth, taskbarHeight, compactGap, comfortableGap, providerTop,
                providerGap, codexMarginLeft, claudeMarginTop, effectiveLayout, taskbar);
        }

        var (width, height) = CalculateWindowSize(
            request, providerCount, small, comfortable, singleRow, themed, providerGap);
        return new WidgetLayoutMetrics(width, height, compactGap, comfortableGap, providerTop, providerGap,
            codexMarginLeft, claudeMarginTop, effectiveLayout);
    }

    /// <summary>
    /// Sizes the in-taskbar overlay: three Claude cells separated by the cell
    /// gap, one Codex cell after the provider gap, and one message cell when no
    /// provider is visible. Hidden providers contribute neither a cell nor a gap.
    /// The progress row is dropped when the band cannot keep one DIP above and
    /// below it, so the height fits every band of at least MinimumTaskbarBandHeight.
    /// </summary>
    private static (double Width, double Height, TaskbarLayoutMetrics Metrics) CalculateTaskbarSize(
        WidgetLayoutRequest request)
    {
        var claudeCells = request.ShowClaude ? 3 : 0;
        var codexCells = request.ShowCodex ? 1 : 0;
        var cells = claudeCells + codexCells;
        var contentWidth = cells == 0
            ? TaskbarCellWidth
            : cells * TaskbarCellWidth + Math.Max(0, claudeCells - 1) * TaskbarCellGap +
              (request.ShowClaude && request.ShowCodex ? TaskbarProviderGap : 0);
        var width = 2 * TaskbarPaddingX + contentWidth;

        var baseHeight = TaskbarLabelLineHeight + TaskbarValueLineHeight + 2 * TaskbarPaddingY +
                         TaskbarClippingGuard;
        var heightWithProgress = baseHeight + TaskbarProgressTopMargin + TaskbarProgressHeight;
        var progressVisible = request.ShowProgressBars && cells > 0 &&
                              request.TaskbarBandHeight >= heightWithProgress + 2;
        var metrics = new TaskbarLayoutMetrics(
            TaskbarCellWidth,
            TaskbarCellGap,
            TaskbarProviderGap,
            TaskbarPaddingX,
            TaskbarPaddingY,
            TaskbarLabelLineHeight,
            TaskbarValueLineHeight,
            TaskbarProgressHeight,
            TaskbarProgressTopMargin,
            TaskbarProgressSideMargin,
            progressVisible,
            TaskbarAnchorGap);
        return (width, progressVisible ? heightWithProgress : baseHeight, metrics);
    }

    private static (double Width, double Height) CalculateWindowSize(WidgetLayoutRequest request,
        int providerCount, bool small, bool comfortable, bool singleRow, bool themed, double providerGap)
    {
        if (providerCount == 0)
            return (small ? 250 : comfortable ? 360 : 300, small ? 34 : comfortable ? 48 : 40);

        if (small)
        {
            var width = request.ShowClaude ? 168 : 72;
            if (singleRow && request.ShowClaude && request.ShowCodex) width = 224;
            width -= 10;

            var height = singleRow ? 60 : providerCount == 2 ? 110 : 60;
            if (themed)
            {
                var chromeExtra = ThemeManager.UsesAngularChrome(request.Theme) ? 2
                    : request.Theme is WidgetVisualTheme.FluentGlass or WidgetVisualTheme.Orbit ? 4 : 1;
                width += chromeExtra;
                height += chromeExtra;
            }
            // Keep two DIPs beyond the measured content boundary. Modern's
            // two-provider vertical stack previously missed its chrome by two
            // additional DIPs because it has no themed chrome allowance.
            height += 2;
            if (!themed && !singleRow && providerCount == 2) height += 2;
            return (width, height);
        }

        var vertical = LinearHeightMetrics.For(request.Theme);
        var outerChromeHeight = comfortable ? vertical.ComfortableOuterChromeHeight : vertical.CompactOuterChromeHeight;
        var rowWithoutProgress = comfortable ? vertical.ComfortableRowWithoutProgress : vertical.CompactRowWithoutProgress;
        var progressFootprint = request.ShowProgressBars
            ? vertical.ProgressVisualHeight + (comfortable ? 7 : 5)
            : 0;
        var rowHeight = rowWithoutProgress + progressFootprint;
        const double clippingGuard = 2;

        if (singleRow)
        {
            var baseWidth = comfortable
                ? request.ShowClaude && request.ShowCodex ? 600 : request.ShowClaude ? 456 : 240
                : request.ShowClaude && request.ShowCodex ? 520 : request.ShowClaude ? 378 : 190;
            if (request.ShowCodex) baseWidth -= comfortable ? 40 : 28;

            // Compact and Comfortable previously spent too much horizontal space per metric.
            // Reduce the content baseline by 25%, while preserving theme-specific chrome room.
            var width = Math.Round(baseWidth * 0.75);
            var visibleMetricCount = (request.ShowClaude ? 3 : 0) + (request.ShowCodex ? 1 : 0);
            width += visibleMetricCount * (request.Theme switch
            {
                WidgetVisualTheme.RetroNight => 10,
                WidgetVisualTheme.FluentGlass => 16,
                WidgetVisualTheme.TerminalMono => 14,
                WidgetVisualTheme.Orbit => 18,
                WidgetVisualTheme.PaperInk => 7,
                _ => 0
            });
            var height = outerChromeHeight + rowHeight + clippingGuard;
            return (width, height);
        }

        var baseWidthTwoRows = comfortable
            ? request.ShowClaude ? 480 : 340
            : request.ShowClaude ? 400 : 260;
        if (!request.ShowClaude && request.ShowCodex) baseWidthTwoRows -= comfortable ? 48 : 36;
        var widthTwoRows = Math.Round(baseWidthTwoRows * 0.75);
        widthTwoRows += request.Theme switch
        {
            WidgetVisualTheme.RetroNight => 20,
            WidgetVisualTheme.FluentGlass => 42,
            WidgetVisualTheme.TerminalMono => 32,
            WidgetVisualTheme.Orbit => 48,
            WidgetVisualTheme.PaperInk => 18,
            _ => 0
        };
        var heightTwoRows = outerChromeHeight + rowHeight * providerCount +
                            Math.Max(0, providerCount - 1) * providerGap + clippingGuard;
        return (widthTwoRows, heightTwoRows);
    }

    /// <summary>
    /// Vertical geometry shared by every linear widget layout. Values include
    /// the root chrome and the still-visible metric card/text row, while the
    /// progress visual is kept separate so hiding it removes only its real
    /// footprint. The two-DIP guard in CalculateWindowSize absorbs layout
    /// rounding at fractional display scales.
    /// </summary>
    private readonly record struct LinearHeightMetrics(
        double CompactOuterChromeHeight,
        double ComfortableOuterChromeHeight,
        double CompactRowWithoutProgress,
        double ComfortableRowWithoutProgress,
        double ProgressVisualHeight)
    {
        internal static LinearHeightMetrics For(WidgetVisualTheme theme) => theme switch
        {
            WidgetVisualTheme.RetroNight => new(20, 25, 25, 30, 8),
            WidgetVisualTheme.FluentGlass => new(18, 23, 31, 36, 6),
            WidgetVisualTheme.TerminalMono => new(18, 23, 28, 33, 16),
            WidgetVisualTheme.Orbit => new(18, 23, 34, 40, 32),
            WidgetVisualTheme.PaperInk => new(18, 23, 24, 28, 12),
            _ => new(20, 26, 16, 17, 4)
        };
    }
}

File: macos/Packages/DejavuKit/Sources/DejavuDomain/WidgetLayoutCalculator.swift
import Foundation

public struct WidgetSize: Hashable, Sendable {
    public let width: Double
    public let height: Double

    public init(width: Double, height: Double) {
        self.width = width
        self.height = height
    }
}

public struct WidgetLayoutRequest: Hashable, Sendable {
    public let density: WidgetDensity
    public let requestedLayout: WidgetLayout
    public let showClaude: Bool
    public let showCodex: Bool
    public let showProgress: Bool
    public let claudeMetricSlots: Int
    public let codexMetricSlots: Int
    public let appearance: ThemePreference
    public let accessibilityTextScale: Double

    public init(
        density: WidgetDensity,
        requestedLayout: WidgetLayout,
        showClaude: Bool,
        showCodex: Bool,
        showProgress: Bool,
        claudeMetricSlots: Int = 2,
        codexMetricSlots: Int = 2,
        appearance: ThemePreference = .system,
        accessibilityTextScale: Double = 1
    ) {
        self.density = density
        self.requestedLayout = requestedLayout
        self.showClaude = showClaude
        self.showCodex = showCodex
        self.showProgress = showProgress
        self.claudeMetricSlots = claudeMetricSlots
        self.codexMetricSlots = codexMetricSlots
        self.appearance = appearance
        self.accessibilityTextScale = accessibilityTextScale
    }

    public init(
        density: WidgetDensity,
        requestedLayout: WidgetLayout,
        serviceState: WidgetServiceState,
        showProgress: Bool,
        appearance: ThemePreference = .system,
        accessibilityTextScale: Double = 1
    ) {
        let visibility = serviceState.visibility
        self.init(
            density: density,
            requestedLayout: requestedLayout,
            showClaude: visibility.showClaude,
            showCodex: visibility.showCodex,
            showProgress: showProgress,
            appearance: appearance,
            accessibilityTextScale: accessibilityTextScale
        )
    }
}

public struct WidgetLayoutMetrics: Hashable, Sendable {
    public let panelSize: WidgetSize
    public let effectiveLayout: WidgetLayout
    public let providerOrder: [UsageProviderID]
    public let providerGap: Double
    public let metricGap: Double
    public let progressFootprint: Double
    public let metricCellWidth: Double
    public let claudeContentWidth: Double
    public let codexContentWidth: Double
    public let smallRingDiameter: Double
    public let contentPadding: Double

    public init(
        panelSize: WidgetSize,
        effectiveLayout: WidgetLayout,
        providerOrder: [UsageProviderID],
        providerGap: Double,
        metricGap: Double,
        progressFootprint: Double,
        metricCellWidth: Double,
        claudeContentWidth: Double,
        codexContentWidth: Double,
        smallRingDiameter: Double,
        contentPadding: Double
    ) {
        self.panelSize = panelSize
        self.effectiveLayout = effectiveLayout
        self.providerOrder = providerOrder
        self.providerGap = providerGap
        self.metricGap = metricGap
        self.progressFootprint = progressFootprint
        self.metricCellWidth = metricCellWidth
        self.claudeContentWidth = claudeContentWidth
        self.codexContentWidth = codexContentWidth
        self.smallRingDiameter = smallRingDiameter
        self.contentPadding = contentPadding
    }

    public var panelWidth: Double { panelSize.width }
    public var panelHeight: Double { panelSize.height }

    public func providerContentWidth(for provider: UsageProviderID) -> Double {
        switch provider {
        case .claude:
            claudeContentWidth
        case .codex:
            codexContentWidth
        }
    }
}

/// AppKit/SwiftUI-free geometry source for every widget service state.
public enum WidgetLayoutCalculator {
    public static func calculate(_ request: WidgetLayoutRequest) -> WidgetLayoutMetrics {
        let providerCount = (request.showClaude ? 1 : 0) + (request.showCodex ? 1 : 0)
        let effectiveLayout: WidgetLayout = request.requestedLayout == .twoRows && providerCount == 2
            ? .twoRows
            : .singleRow
        let scale = normalizedTextScale(request.accessibilityTextScale)

        let baseMetricCellWidth: Double
        let metricGap: Double
        let padding: Double
        let providerGap: Double
        let rowWithoutProgress: Double
        let progressFootprint: Double

        switch request.density {
        case .small:
            baseMetricCellWidth = 48
            metricGap = 6
            padding = 10
            providerGap = providerCount == 2 ? 8 : 0
            rowWithoutProgress = max(48, ceil(40 * scale))
            progressFootprint = 0
        case .compact:
            baseMetricCellWidth = 72
            metricGap = 10
            padding = 12
            providerGap = providerCount == 2 ? 10 : 0
            rowWithoutProgress = ceil(35 * scale)
            progressFootprint = request.showProgress ? 9 : 0
        case .comfortable:
            baseMetricCellWidth = 88
            metricGap = 14
            padding = 14
            providerGap = providerCount == 2 ? 12 : 0
            rowWithoutProgress = ceil(42 * scale)
            progressFootprint = request.showProgress ? 11 : 0
        }

        let metricCellWidth = ceil(baseMetricCellWidth * scale)
        let claudeContentWidth = providerWidth(
            metricCellWidth: metricCellWidth,
            metricGap: metricGap,
            slots: request.claudeMetricSlots
        )
        let codexContentWidth = providerWidth(
            metricCellWidth: metricCellWidth,
            metricGap: metricGap,
            slots: request.codexMetricSlots
        )
        let rowHeight = rowWithoutProgress + progressFootprint
        let size: WidgetSize

        if providerCount == 0 {
            let emptyWidth: Double
            let emptyHeight: Double
            switch request.density {
            case .small:
                emptyWidth = 250
                emptyHeight = 34
            case .compact:
                emptyWidth = 300
                emptyHeight = 40
            case .comfortable:
                emptyWidth = 360
                emptyHeight = 48
            }
            size = WidgetSize(width: ceil(emptyWidth * scale), height: ceil(emptyHeight * scale))
        } else if effectiveLayout == .singleRow {
            let contentWidth = (request.showClaude ? claudeContentWidth : 0)
                + (request.showCodex ? codexContentWidth : 0)
                + providerGap
            size = WidgetSize(
                width: ceil(contentWidth + padding * 2),
                height: ceil(rowHeight + padding * 2)
            )
        } else {
            size = WidgetSize(
                width: ceil(max(claudeContentWidth, codexContentWidth) + padding * 2),
                height: ceil(rowHeight * 2 + providerGap + padding * 2)
            )
        }

        let providerOrder: [UsageProviderID]
        if effectiveLayout == .twoRows {
            providerOrder = [.codex, .claude]
        } else {
            providerOrder = [
                request.showClaude ? .claude : nil,
                request.showCodex ? .codex : nil
            ].compactMap { $0 }
        }

        return WidgetLayoutMetrics(
            panelSize: size,
            effectiveLayout: effectiveLayout,
            providerOrder: providerOrder,
            providerGap: providerGap,
            metricGap: metricGap,
            progressFootprint: progressFootprint,
            metricCellWidth: metricCellWidth,
            claudeContentWidth: claudeContentWidth,
            codexContentWidth: codexContentWidth,
            smallRingDiameter: 30,
            contentPadding: padding
        )
    }

    private static func normalizedTextScale(_ value: Double) -> Double {
        guard value.isFinite else { return 1 }
        return min(2, max(1, value))
    }

    private static func providerWidth(
        metricCellWidth: Double,
        metricGap: Double,
        slots: Int
    ) -> Double {
        let normalizedSlots = min(3, max(1, slots))
        return metricCellWidth * Double(normalizedSlots)
            + metricGap * Double(normalizedSlots - 1)
    }
}
